test/cfa/sudoers_test.rb

```diff
@@ -0,0 +1,183 @@
+        expect(sudoers.users_specs).to eq([])
+      end
+
+      context "but a new user specs were added" do
+        before do
+          sudoers.add_user(yast_user_spec)
+          sudoers.add_user(cli_user_spec)
+        end
+
+        it "returns a collection holding all user specs" do
+          expect(sudoers.users_specs).to eq([yast_user_spec, cli_user_spec])
+        end
+      end
+    end
+  end
+
+  describe "#clean_users" do
+    before { sudoers.load }
+
+    it "removes all users specifications" do
+      initial_users_specs = sudoers.users_specs
+
+      sudoers.clean_users
+
+      expect(sudoers.users_specs).to_not eq(initial_users_specs)
+      expect(sudoers.users_specs).to eq([])
+    end
+  end
+
+  describe "#save" do
+    let(:file_path) { File.join(tmpdir, "sudoers") }
+    let(:tmpdir) { Dir.mktmpdir }
+
+    before do
+      # Use a copy of the example file
+      FileUtils.cp(example_file, file_path)
+
+      sudoers.load
+    end
+
+    after do
+      FileUtils.remove_entry(tmpdir)
+    end
+
+    context "when the file does not contain user specifications" do
+      let(:example_file) { File.join(DATA_PATH, "sudoers_no_specs_example") }
+
+      context "nor does it include the #includedir sudoers.d directive" do
+        let(:example_file) { File.join(DATA_PATH, "sudoers_old_example") }
+
+        it "writes new specifications at the end of file" do
+          sudoers.add_user(yast_user_spec)
+          sudoers.add_user(cli_user_spec)
+          sudoers.save
+
+          expect(File.read(file_path)).to match(/.*cli ALL = \(ALL\) \/sbin\/yast\Z/)
+        end
+      end
+
+      context "but does include the #includedir sudoers.d directive" do
+        it "writes new specifications before it" do
+          sudoers.add_user(yast_user_spec)
+          sudoers.add_user(cli_user_spec)
+          sudoers.save
+
+          expect(File.read(file_path)).to match(/.*yast.*cli.*\#includedir.*sudoers\.d.*/m)
+        end
+      end
+    end
+
+    context "when the file already contains user specs" do
+      it "writes new specifications after them" do
+        sudoers.add_user(yast_user_spec)
+        sudoers.add_user(cli_user_spec)
+        sudoers.save
+
+        expect(File.read(file_path)).to include(
+          "root ALL=(ALL) ALL\n" \
+          "yast ALL = (root) NOPASSWD : /sbin/yast2\n" \
+          "cli ALL = (ALL) /sbin/yast\n"
+        )
+      end
+
+      context "but they are cleaned before start adding new ones" do
+        it "writes new specs starting in the previous ones position" do
+          previous_first_position = File.read(file_path) =~ /.*^ALL.*ALL.*=\(ALL\)/
+
+          sudoers.clean_users
+          sudoers.add_user(yast_user_spec)
+          sudoers.add_user(cli_user_spec)
+          sudoers.save
+
+          new_first_position = File.read(file_path) =~ /.*^yast.*ALL.*=/
+
+          expect(new_first_position).to eq(previous_first_position)
+        end
+
+      end
+    end
+  end
```

Comment:
so what is not tested here:

1. case with collection fixes ( so single user )
2. empty commands and maybe other cases as no option is specified as mandatory

maybe more, but I do not see it now :)

Reply:
> case with collection fixes ( so single user )

To avoid bloating the data directory with new examples, I am going to comment out the "Defaults targetpw" lines in the _test/data/sudoers_example_ file, making that test implicitely. Do you think is enough or do you prefer an explicit test?

> empty commands and maybe other cases as no option is specified as mandatory

I added a couple of `raise ArgumentError` and their correspondents tests.

Reply:
well, for single user I expect explicit test where is just one user specified in file. So collection fix can take effect.